File: totoggle_go/metrics_test.go
package totoggle

import (
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
)

type recordingSuccessListener struct{ counts []int }

func (r *recordingSuccessListener) OnRefreshSuccess(toggleCount int) {
	r.counts = append(r.counts, toggleCount)
}

type recordingFailureListener struct {
	errs     []error
	failures []int
}

func (r *recordingFailureListener) OnRefreshFailure(err error, consecutiveFailures int) {
	r.errs = append(r.errs, err)
	r.failures = append(r.failures, consecutiveFailures)
}

type recordingEvaluationListener struct {
	paths   []string
	results []bool
}

func (r *recordingEvaluationListener) OnEvaluation(path string, result bool) {
	r.paths = append(r.paths, path)
	r.results = append(r.results, result)
}

// A listener implementing all three interfaces at once must be dispatched by every one of them.
type allThreeListener struct {
	recordingSuccessListener
	recordingFailureListener
	recordingEvaluationListener
}

func TestMetricsRegistry_AddMetricsListener_RegistersOnlyImplementedInterfaces(t *testing.T) {
	reg := newMetricsRegistry()
	success := &recordingSuccessListener{}
	reg.add(success)

	reg.notifyRefreshSuccess(3)
	reg.notifyRefreshFailure(errors.New("boom"), 1)
	reg.notifyEvaluation("t1", true)

	assert.Equal(t, []int{3}, success.counts)
}

func TestMetricsRegistry_DispatchesToAllThreeWhenAllImplemented(t *testing.T) {
	reg := newMetricsRegistry()
	l := &allThreeListener{}
	reg.add(l)

	reg.notifyRefreshSuccess(5)
	reg.notifyRefreshFailure(errors.New("boom"), 2)
	reg.notifyEvaluation("t1.t2", false)

	assert.Equal(t, []int{5}, l.recordingSuccessListener.counts)
	assert.Equal(t, []int{2}, l.recordingFailureListener.failures)
	assert.Equal(t, []string{"t1.t2"}, l.recordingEvaluationListener.paths)
	assert.Equal(t, []bool{false}, l.recordingEvaluationListener.results)
}

type panickingListener struct{}

func (panickingListener) OnEvaluation(path string, result bool) { panic("listener bug") }

// A broken listener must never take down evaluation or the refresh loop.
func TestMetricsRegistry_PanickingListenerDoesNotPropagate(t *testing.T) {
	reg := newMetricsRegistry()
	reg.add(panickingListener{})
	good := &recordingEvaluationListener{}
	reg.add(good)

	assert.NotPanics(t, func() {
		reg.notifyEvaluation("t1", true)
	})
	assert.Equal(t, []string{"t1"}, good.paths)
}

func TestMetricsRegistry_UnrelatedTypeRegistersForNothing(t *testing.T) {
	reg := newMetricsRegistry()
	reg.add("not a listener")

	assert.NotPanics(t, func() {
		reg.notifyRefreshSuccess(1)
		reg.notifyRefreshFailure(errors.New("x"), 1)
		reg.notifyEvaluation("t1", true)
	})
}

File: totoggle_node/src/metrics.test.ts
import { describe, expect, it } from "vitest";
import { MetricsDispatcher } from "./metrics.js";
import type { ToToggleMetricsListener } from "./metrics.js";

describe("MetricsDispatcher", () => {
  it("calls only the methods a partial listener implements", () => {
    const dispatcher = new MetricsDispatcher();
    const successCounts: number[] = [];
    const listener: ToToggleMetricsListener = {
      onRefreshSuccess: (count) => successCounts.push(count),
    };
    dispatcher.add(listener);

    dispatcher.notifyRefreshSuccess(3);
    dispatcher.notifyRefreshFailure(new Error("boom"), 1);
    dispatcher.notifyEvaluation("t1", true);

    expect(successCounts).toEqual([3]);
  });

  it("dispatches to all three methods when a listener implements all three", () => {
    const dispatcher = new MetricsDispatcher();
    const calls: string[] = [];
    const listener: ToToggleMetricsListener = {
      onRefreshSuccess: (count) => calls.push(`success:${count}`),
      onRefreshFailure: (err, failures) => calls.push(`failure:${err.message}:${failures}`),
      onEvaluation: (path, result) => calls.push(`eval:${path}:${result}`),
    };
    dispatcher.add(listener);

    dispatcher.notifyRefreshSuccess(5);
    dispatcher.notifyRefreshFailure(new Error("boom"), 2);
    dispatcher.notifyEvaluation("t1.t2", false);

    expect(calls).toEqual(["success:5", "failure:boom:2", "eval:t1.t2:false"]);
  });

  // A broken listener must never take down evaluation or the refresh loop.
  it("a throwing listener does not propagate or block later listeners", () => {
    const dispatcher = new MetricsDispatcher();
    const calls: string[] = [];
    dispatcher.add({
      onEvaluation: () => {
        throw new Error("listener bug");
      },
    });
    dispatcher.add({ onEvaluation: (path) => calls.push(path) });

    expect(() => dispatcher.notifyEvaluation("t1", true)).not.toThrow();
    expect(calls).toEqual(["t1"]);
  });

  it("an empty listener object registers for nothing and never throws", () => {
    const dispatcher = new MetricsDispatcher();
    dispatcher.add({});

    expect(() => {
      dispatcher.notifyRefreshSuccess(1);
      dispatcher.notifyRefreshFailure(new Error("x"), 1);
      dispatcher.notifyEvaluation("t1", true);
    }).not.toThrow();
  });
});
